tests_tf/test_attacks.py

```diff
@@ -44,6 +57,175 @@ def test_generate_np(self):
         # test perturbation norm
         self.assertTrue(np.allclose(perturbation_norm, self.attack.eps))
 
+class TestFastGradientMethod(CleverHansTest):
+    def setUp(self):
+        super(TestFastGradientMethod, self).setUp()
+        import tensorflow as tf
+
+        # The world's simplest neural network
+        def my_model(x):
+            W1 = tf.constant([[1.5, .3], [-2, 0.3]], dtype=tf.float32)
+            h1 = tf.nn.sigmoid(tf.matmul(x, W1))
+            W2 = tf.constant([[-2.4, 1.2], [0.5, -2.3]], dtype=tf.float32)
+            res = tf.nn.softmax(tf.matmul(x, W2))
+            return res
+
+        self.sess = tf.Session()
+        self.model = my_model
+        self.attack = FastGradientMethod(self.model, sess=self.sess)
+
+    def test_generate_np_gives_adversarial_example(self):
+        x_val = np.random.rand(100, 2)
+        x_val = np.array(x_val, dtype=np.float32)
+
+        for ord in [np.inf, 1, 2]:
+            x_adv = self.attack.generate_np(x_val, eps=.5, ord=ord,
+                                            clip_min=-5, clip_max=5)
+
+            if ord == np.inf:
+                assert np.allclose(np.max(np.abs(x_adv-x_val), axis=1), 0.5)
+            elif ord == 1:
+                assert np.allclose(np.sum(np.abs(x_adv-x_val), axis=1), 0.5)
+            elif ord == 2:
+                assert np.allclose(np.sum(np.square(x_adv-x_val), axis=1)**.5,
+                                   0.5)
+
+            orig_labs = np.argmax(self.sess.run(self.model(x_val)), axis=1)
+            new_labs = np.argmax(self.sess.run(self.model(x_adv)), axis=1)
+
+            assert np.mean(orig_labs==new_labs) < 0.5
+
+    def test_generate_np_can_be_called_with_different_eps(self):
+        x_val = np.random.rand(100, 2)
```

Comment:
For some tests, they could be flaky. This one in particular should never fail since we're not checking we get adversarial examples back, just that the distance is right.